onnxruntime/contrib_ops/rocm/bert/multihead_attention.cu

```diff
@@ -0,0 +1,143 @@
+namespace contrib {
+namespace rocm {
+
+#define REGISTER_KERNEL_TYPED(T)                                  \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                  \
+      MultiHeadAttention,                                         \
+      kMSDomain,                                                  \
+      1,                                                          \
+      T,                                                          \
+      kRocmExecutionProvider,                                     \
+      (*KernelDefBuilder::Create())                               \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
+      MultiHeadAttention<T>);
+
+REGISTER_KERNEL_TYPED(float)
+REGISTER_KERNEL_TYPED(MLFloat16)
+
+template <typename T>
+MultiHeadAttention<T>::MultiHeadAttention(const OpKernelInfo& info)
+    : RocmKernel(info),
+      cumulated_sequence_length_q_cache_(),
+      cumulated_sequence_length_kv_cache_() {
+  int64_t num_heads = 0;
+  ORT_ENFORCE(info.GetAttr("num_heads", &num_heads).IsOK() && num_heads > 0);
+  num_heads_ = static_cast<int>(num_heads);
+
+  mask_filter_value_ = info.GetAttrOrDefault<float>("mask_filter_value", -10000.0f);
+
+  scale_ = info.GetAttrOrDefault<float>("scale", 0.0f);
+
+  disable_fused_self_attention_ =
+      sizeof(T) != 2 ||
+      ParseEnvironmentVariableWithDefault<bool>(attention::kDisableFusedSelfAttention, false);
+
+  disable_fused_cross_attention_ =
+      sizeof(T) != 2 ||
+      ParseEnvironmentVariableWithDefault<bool>(attention::kDisableFusedCrossAttention, false);
+
+  // Allocate cache buffers
+  constexpr size_t cache_bytes = sizeof(int32_t) * (static_cast<size_t>(kCumulatedSequenceLengthCacheMaxBatchSize) + 1);
+  cumulated_sequence_length_q_cache_.buffer = GetTransientScratchBuffer<void>(cache_bytes);
+  cumulated_sequence_length_q_cache_.max_batch_size = kCumulatedSequenceLengthCacheMaxBatchSize;
+  cumulated_sequence_length_kv_cache_.buffer = GetTransientScratchBuffer<void>(cache_bytes);
+  cumulated_sequence_length_kv_cache_.max_batch_size = kCumulatedSequenceLengthCacheMaxBatchSize;
+}
+
+template <typename T>
+Status MultiHeadAttention<T>::ComputeInternal(OpKernelContext* context) const {
+  ORT_ENFORCE(
+      GetTuningContext()->IsTunableOpEnabled(),
+      "MultiHeadAttention of ROCm EP is only supported if tunable op is used and tuning is enabled.");
+
+  const Tensor* query = context->Input<Tensor>(0);
+  const Tensor* key = context->Input<Tensor>(1);
+  const Tensor* value = context->Input<Tensor>(2);
+  const Tensor* bias = context->Input<Tensor>(3);
+  const Tensor* key_padding_mask = context->Input<Tensor>(4);
+  const Tensor* relative_position_bias = context->Input<Tensor>(5);
+  const Tensor* past_key = context->Input<Tensor>(6);
+  const Tensor* past_value = context->Input<Tensor>(7);
+
+  // TODO: relax
+  ORT_ENFORCE(bias == nullptr && key_padding_mask == nullptr && past_key == nullptr && past_value == nullptr);
```

Comment:
> // TODO: relax

Is this still a TODO? If so, please be more descriptive about what needs to be relaxed here.

Reply:
Yes, it is. The following `ORT_ENFORCE` indicate we don't support many things now. These constraints needs to be relaxed in the future.

Reply:
Updated.